### PR TITLE
Fix/84 unsave password handling

### LINKED ISSUE #84: unsafe password handling
actual plaintext passwords being sent in user management, really unsafe thing to do. Standard way is to salt the passwords and the compare the hashes, not plain text passwords

### PR DESCRIPTION
**Pull request by ledex**
Add hashing and salting of passwords in the database with [bcrypt](https://godoc.org/golang.org/x/crypto/bcrypt). This fixes #84. 

Added that to the next minor version, because changes to the database have to be made and old users have to be transferred.

### CODE REVIEW

**Comment by ledex** [comment]
Currently, the password is still represented as a string. This might not be wanted and instead, the password should be represented as binary.

File: modules/userman/email.go [hunk 1]
```diff
@@ -49,8 +51,17 @@ func (m *Module) HandleEmailSignIn() http.HandlerFunc {
...
+		err = bcrypt.CompareHashAndPassword([]byte(userObj["pass"].(string)), []byte(req["pass"].(string)))
+		if err != nil {
+			w.WriteHeader(http.StatusUnauthorized)
+			json.NewEncoder(w).Encode(map[string]string{"error": "Given credentials are not correct"})
+			return
+		}
+
+		// Delete password from user
 		delete(userObj, "pass")
 		delete(req, "pass")
```
Comment (Jayesh333) [inline modules/userman/email.go]: No need to delete `pass` from request

File: modules/userman/email.go [hunk 2]
```diff
@@ -132,3 +146,18 @@ func (m *Module) HandleEmailSignUp() http.HandlerFunc {
 		json.NewEncoder(w).Encode(map[string]interface{}{"user": req, "token": token})
 	}
 }
+
+func hashPassword(pwd string) string {
+	//Generates a new hash from the given password
+	hash, err := bcrypt.GenerateFromPassword([]byte(pwd), bcrypt.MinCost)
+	if err != nil {
+		panic(err.Error)
```
Comment (Jayesh333) [inline modules/userman/email.go]: Can we return an InternalServerError instead of panic?

File: modules/userman/email.go [hunk 3]
```diff
@@ -132,3 +146,18 @@ func (m *Module) HandleEmailSignUp() http.HandlerFunc {
...
+	}
+
+	//Checks if the hash is correct for the given password
+	err = bcrypt.CompareHashAndPassword(hash, []byte(pwd))
+	if err != nil {
+		panic(err.Error)
```
Comment (Jayesh333) [inline modules/userman/email.go]: Return an error rather than panic

File: modules/userman/email.go (hunk 2 above)
Comment (ledex) [reply modules/userman/email.go]: Sure, but I would suggest doing both so that the administrator can see in the logs that an error was raised.

**Comment by ledex** [comment]
> Need to add `id` / `_id` (for SQL and MongoDb respectively) in the `req` object from the `userObj` before creating `token` in the signin as it is done for signup

Done 👍

File: modules/userman/email.go [hunk 4]
```diff
@@ -49,12 +51,26 @@ func (m *Module) HandleEmailSignIn() http.HandlerFunc {
...
+			return
+		}
+
+		// Delete password from user
 		delete(userObj, "pass")
 		delete(req, "pass")
 
 		// Create a token
+		if dbType == "mongo" {
+			req["_id"] = userObj["_id"]
```
Comment (YourTechBud) [inline modules/userman/email.go]:
```suggestion
			req["id"] = userObj["_id"]
```

File: modules/userman/email.go [hunk 5]
```diff
@@ -49,12 +51,26 @@ func (m *Module) HandleEmailSignIn() http.HandlerFunc {
...
+			json.NewEncoder(w).Encode(map[string]string{"error": "Given credentials are not correct"})
+			return
+		}
+
+		// Delete password from user
 		delete(userObj, "pass")
 		delete(req, "pass")
 
 		// Create a token
+		if dbType == "mongo" {
```
Comment (YourTechBud) [inline modules/userman/email.go]:
```suggestion
		if dbType == string(utils.Mongo) {
```

File: modules/userman/email.go [hunk 6]
```diff
@@ -49,12 +51,26 @@ func (m *Module) HandleEmailSignIn() http.HandlerFunc {
...
+			req["_id"] = userObj["_id"]
+		} else {
+			req["id"] = userObj["id"]
+		}
 		req["role"] = userObj["role"]
 		req["name"] = userObj["name"]
```
Comment (YourTechBud) [inline modules/userman/email.go]: We don't need the name to be there in the token. Will increase the token size.

File: modules/userman/email.go [hunk 7]
```diff
@@ -94,9 +110,19 @@ func (m *Module) HandleEmailSignUp() http.HandlerFunc {
...
+		req["pass"], err = hashPassword(req["pass"].(string))
+		if err != nil {
+			log.Println("Err: ", err)
+			w.WriteHeader(http.StatusInternalServerError)
+			json.NewEncoder(w).Encode(map[string]string{"error": "Failed to hash password"})
+			return
+		}
+
 		// Create read request
 		readReq := &model.ReadRequest{Find: map[string]interface{}{"email": req["email"], "pass": req["pass"]}, Operation: utils.One}
```
Comment (YourTechBud) [inline modules/userman/email.go]:
```suggestion
		readReq := &model.ReadRequest{Find: map[string]interface{}{"email": req["email"]}, Operation: utils.One}
```

Comment (YourTechBud) [reply modules/userman/email.go]: Can't keep the pass here since it will be different every time.

**Comment by ledex** [comment]
Okey. Should be good to go... Hopefully 😄

**Pull request merged**